### ISSUE TITLE
Peformance concerns

### ISSUE THREAD

**Issue by RyanTheAllmighty** [driesvints/blade-icons#71]
Hi,

I updated yesterday from the `nothingworks/blade-svg` package to the `blade-ui-kit/blade-icons` package and I'm seeing much higher CPU usage on my servers as well as a very large number of php slow log entries relating to the package on every single request.

I am using the [owenvoke/blade-fontawesome](https://github.com/owenvoke/blade-fontawesome) add on.

It seems when there is a large number of icons, for instance the fontawesome package provides around 1,500 icons, it loads every single file in the filesystem on every request.

Looking at https://github.com/blade-ui-kit/blade-icons/blob/main/src/Factory.php#L68-L81 it loads all the files in the filesystem (whether or not they're used) on every page request:

But the previous package would only load them as they're used https://github.com/blade-ui-kit/blade-icons/blob/v0.3.4/src/SvgFactory.php#L105

As an example of the slow log I'm seeing on every request:

```
[07-Aug-2020 02:32:07]  [pool www] pid 8233
script_filename = /srv/public/index.php
[0x00007f5fa7616e40] hasChildren() /srv/vendor/symfony/finder/Iterator/ExcludeDirectoryFilterIterator.php:76
[0x00007f5fa7616dd0] hasChildren() /srv/vendor/symfony/finder/Iterator/SortableIterator.php:91
[0x00007f5fa7616d80] next() /srv/vendor/symfony/finder/Iterator/SortableIterator.php:91
[0x00007f5fa7616d30] next() /srv/vendor/symfony/finder/Iterator/SortableIterator.php:91
[0x00007f5fa7616cc0] iterator_to_array() /srv/vendor/symfony/finder/Iterator/SortableIterator.php:91
[0x00007f5fa7616c40] getIterator() /srv/vendor/symfony/finder/Finder.php:772
[0x00007f5fa7616b10] searchInDirectory() /srv/vendor/symfony/finder/Finder.php:614
[0x00007f5fa7616a60] getIterator() /srv/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php:460
[0x00007f5fa76169f0] iterator_to_array() /srv/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php:460
[0x00007f5fa7616970] allFiles() /srv/vendor/blade-ui-kit/blade-icons/src/Factory.php:71
[0x00007f5fa76168b0] registerComponents() /srv/vendor/blade-ui-kit/blade-icons/src/BladeIconsServiceProvider.php:49
[0x00007f5fa7616830] BladeUI\Icons\{closure}() /srv/vendor/laravel/framework/src/Illuminate/Container/Container.php:1141
[0x00007f5fa7616790] fireCallbackArray() /srv/vendor/laravel/framework/src/Illuminate/Container/Container.php:1105
[0x00007f5fa7616710] fireAfterResolvingCallbacks() /srv/vendor/laravel/framework/src/Illuminate/Container/Container.php:1090
[0x00007f5fa7616690] fireResolvingCallbacks() /srv/vendor/laravel/framework/src/Illuminate/Container/Container.php:711
[0x00007f5fa76165b0] resolve() /srv/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:796
[0x00007f5fa7616510] resolve() /srv/vendor/laravel/framework/src/Illuminate/Container/Container.php:637
[0x00007f5fa7616490] make() /srv/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:781
[0x00007f5fa7616400] make() /srv/vendor/laravel/framework/src/Illuminate/Container/Container.php:1284
[0x00007f5fa7616390] offsetGet() unknown:0
```

I've had to revert this package upgrade and move back to the deprecated old package which immediately solved the slowlog and cpu issues.

**Comment by driesvints** [comment]
Yes, the problem with the new version is that we need the icon names up front to register the components. I'll think about a new way to do this, maybe with some sort of caching system. Perhaps through a command which can be run. 

Not sure when I'll get to this but appreciating any help in the meantime. Thanks for reporting.

**Comment by driesvints** [comment]
@RyanTheAllmighty one thing I'm wondering: does most of the performance hits come from reading the file or the registration of the components?

**Comment by RyanTheAllmighty** [comment]
@driesvints looking at the slow logs, it's always the `allFiles` call that is taking more than 5 seconds.

I thought it might be that all the file reading on every request is tying up all the disk IO on the server, but looking at the server stats, that's not the case.

I did some basic debug logging in a testing environment, and the `allFiles` call is taking up a majority of the time and registering the components only taking about 10% of the time.

For instance with the largest fontawesome set of 1,000 icons taking around 20-25ms for `allFiles` and around 2ms for the registering of all the components.

It might just be a case of with smaller number of requests coming in, it's fine and can handle it, but once you have any larger number of requests come in, it starts to take an effect as it's constantly rereading all the files in and registering all the components.

**Comment by driesvints** [comment]
@RyanTheAllmighty gotcha. Thanks for all your investigation here 👍 

I think there's two things we can do here:

1. Cache the component names in a separate file similar to the `packages.php` from Laravel. That way we only need to read one file instead of using `allFiles` all the time
2. Provide a way for users to manually build that list so the library only knows about the listed icons and not every icon from a set.

I think 1. is a given that we should implement with 2. being a nice to have. I'll try to check into this later on but at the moment I'm a bit busy getting the Blade UI Kit release out of the door with a couple of other things after that. If anyone wants to help out here before I get to it I'll definitely appreciate it.

**Comment by ragulka** [comment]
I was using blade-svg before and built my own blade icon component around it. What I did was to simply have the icon name provided as an attribute:
```
<x-icon name="chevron-down" />
```
Wouldn't a similar approach work?

**Comment by driesvints** [comment]
@ragulka no because then you'll lose the benefit of a much nicer api to reference the icons.

**Comment by ragulka** [comment]
I guess that's a matter of taste. I actually prefer typing out the attribute. It also has the benefit of being able to provide the icon name dynamically:

```php
<x-icon :name="$iconName" />
```

It's quite useful when generating a bunch of HTML, like table headers, where you have an array for input:
```
$headers = [
    [
        'title' => 'Name',
        'icon' => 'user'
    ],
    [
        'title' => 'Company',
        'icon' => 'briefcase'
    ]
]

@foreach($headers as $header)
    <tr>
        <th><x-icon :name="$header['icon']" /> {{ $title }}</th>
    </tr>
@endforeach
```

As far as I understand, this is not possible with the current API (except for when using the blade directive)?

**Comment by driesvints** [comment]
@ragulka you can still use the directive for that. It works exactly the same. I don't have any plans for implementing the above, sorry.

**Comment by driesvints** [comment]
@ragulka actually, I've changed my mind. I'll try looking into shipping a default `<x-icon` in the next 0.5 release.

**Comment by owenvoke** [comment]
@driesvints, will that be in addition to the `x-icon-${name}` syntax? Or will `<x-icon :name="$foo">` be the new syntax?

**Comment by driesvints** [comment]
@owenvoke in addition. Current way of working will stay.

**Comment by joshhanley** [comment]
Hi Dries,

Really enjoying the package and the fact that we can pull in different icon sets easily!

I'm also experiencing performance issues after I pulled in [Blade Material Design Icons](https://github.com/renoki-co/blade-mdi) add on.

Took me a little while yesterday to realise what it was, but running my test suite normally takes ~30secs but after pulling in the above package, it's now taking ~1min40secs.

For reference, there is 5,536 icons in that package.

Do you require any more information other than what @RyanTheAllmighty has already provided?
If so I'm happy to run some tests for you.

**Comment by thewebartisan7** [comment]
I would like to know also about using components for icons even as suggested with `<x-icon name="icon-class">`, doesn't mean this that for each icons there is an inclusion of component file? I am often concerned to not load too many views for example for layout like head, navigation, sidebar, footer, etc, and try to put everything in main layout app.blade.php for avoid too many views inclusion. 

I was surprise to see a component for just an icon. Was not better a blade directive in this case? I see there is also directive, but I am asking why make also components.
As I understand for each component, even if are the same, there is a file inclusion. 

Thanks

**Comment by driesvints** [comment]
@thewebartisan7 I much prefer the component syntax over the directive. It's there so people can choose what they want to use. I'm still planning on introducing an `<x-icon name="">` style component. Haven't gotten to that yet.

**Comment by thewebartisan7** [comment]
I see that you are working on single component with name prop. However, what I asked was even using single component <x-icon/> if a page have 10 icons or more, this mean that for each icons there is a templates rendered of <x-icon> component?

I understand that you prefer more the component syntax, I also think is better. But I am concerned about too many views templates being rendered in a page.

For example I often use icon question in a form for each fields to show info in tooltip or inline, then menu have icons, etc.

**Comment by driesvints** [comment]
I'll try to look into all of this in the near future. Just haven't gotten to it yet with the Blade UI Kit launch and all. There's already a PR open that looks promising.

**Comment by rennokki** [comment]
https://github.com/blade-ui-kit/blade-icons/pull/85 should do the job by caching them.

**Comment by anibalsanchez** [comment]
Same issue here using `owenvoke/blade-fontawesome` , the page load was being delayed a whole second due to icon processing.

I solved the issue moving the 23 icons that we use to a `resources/svg` directory and removing the dependency to the whole set of icons from `owenvoke/blade-fontawesome`.

At this time, I think that it is better if a note for **performance tunning** is added to the README.

**Comment by driesvints** [comment]
Hi there. I'm sorry it took me this long to get to this. I'm happy to say that the next version of Blade Icons will include icon caching that will solve the existing performance problems. I'm currently looking for testers of the caching feature and would appreciate feedback on there PR here: https://github.com/blade-ui-kit/blade-icons/pull/121#issuecomment-798906016

**Comment by driesvints** [comment]
@RyanTheAllmighty would very much appreciate it if you could help test this one out :)

**Comment by RyanTheAllmighty** [comment]
> @RyanTheAllmighty would very much appreciate it if you could help test this one out :)

Definitely. Will give it a shot in the next day or two and report back. Will be nice to get off the old depreciated package

**Comment by driesvints** [comment]
Hey @RyanTheAllmighty, did you already had a chance to try it out? The reason I ask is that I want to make sure it fixes things for you before I tag a 1.0.0 release. I also saw a slight increase in CPU locally so want to make sure it's not too big of a deal for you on production. Let me know!

**Comment by RyanTheAllmighty** [comment]
> Hey @RyanTheAllmighty, did you already had a chance to try it out? The reason I ask is that I want to make sure it fixes things for you before I tag a 1.0.0 release. I also saw a slight increase in CPU locally so want to make sure it's not too big of a deal for you on production. Let me know!

Yup. Been an insane week, and with Laracon today, didn't manage to find the time. I have some time tonight to take a look, I'll get back to you on my results in the next couple hours :)

**Comment by RyanTheAllmighty** [comment]
Okay I've made the change again and pushed it to prod. So far seems good, I'm caching the icons as part of the deploy process and verified it's in the bootstrap/cache folder on the server and yeah not seen any issues yet. I'll monitor it overnight and check the slowlogs, but from memory of last time, it was pretty immediate that it happened

**Comment by driesvints** [comment]
Thanks @RyanTheAllmighty!

**Comment by RyanTheAllmighty** [comment]
Looks all good after sitting overnight. I'm fairly confident the issue is fixed now. Thanks a bunch @driesvints should be good to tag a release

**Comment by driesvints** [comment]
@RyanTheAllmighty that's great to hear! I'll be tagging a 1.0.0 version soon.

**Comment by lao9s** [comment]
I tried `php artisan icons:cache`, and the app is loading a little faster and there is not much difference. 

- With cache ~1.4s
- Without cache: ~2s
- Without this package: ~275ms

I noticed that locally is working fast, the problem appears when you deploy the app on a VPS server.

**Comment by vjanssens** [comment]
We've had a lengthy debug session after we've deployed an update to Laravel + Filament v3 and this eventually turned out to be the culprit. As @lao9s described, we could not reproduce it locally but had massive performance issues on a server. We've debugged this using php-fpm slow log.

What's interesting though is that Filament provides their own workaround, which is not documented (https://github.com/filamentphp/filament/blob/e8d44231fcba1a0f4e9625afe7368447d56b2866/packages/panels/src/Http/Middleware/DisableBladeIconComponents.php).

```php
protected $middleware = [
    ...
    \Filament\Http\Middleware\DisableBladeIconComponents::class
];
```

Simply include this middleware and the blade icons are not loaded anymore.
